File: 0023 Merge k Sorted Lists/solution.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.empty())  return NULL;
        int n = lists.size();
        while(n > 1){
            int k = (n + 1) / 2;
            for(int i=0; i<n/2; i++){
                lists[i] = mergeTwoLists(lists[i], lists[i+k]);
            }
            n = k;
        }
        return lists[0];
    }
    
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        
        ListNode* head = new ListNode(-9999);
        ListNode* p = head;
        // head->next = (l1->val >= l2->val) ? l1 : l2;
        while(l1 && l2){
            if(l1->val <= l2->val){
                p->next = l1;
                l1 = l1->next;
                p = p->next;
            }
            else{
                p->next = l2;
                l2 = l2->next;
                p = p->next;
            }
        }
        if(l1)
            p->next = l1;
        if(l2)
            p->next = l2;
        
        return head->next;
    }
};

ListNode *create_linkedlist(std::initializer_list<int> lst)
{
    auto iter = lst.begin();
    ListNode *head = lst.size() ? new ListNode(*iter++) : nullptr;
    for (ListNode *cur=head; iter != lst.end(); cur=cur->next)
        cur->next = new ListNode(*iter++);
    return head;
}

void clear(ListNode *head)
{
    while (head) {
        ListNode *del = head;
        head = head->next;
        delete del;
    }
}

void printList(ListNode *head) {
    for (ListNode *cur = head; cur; cur = cur->next)
        cout << cur->val << "->";
    cout << "\b\b  " << endl;
}

int main()
{
    Solution s;
    vector<ListNode *> vec{
        create_linkedlist({1,4,5}),
        create_linkedlist({1,3,4}),
        create_linkedlist({2,6})
    };
    ListNode *ret = s.mergeKLists(vec);
    printList(ret);
    clear(ret);

    return 0;
}

File: 0023 Merge k Sorted Lists/solution.py
#! python3

from queue import PriorityQueue

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeKLists(self, lists):
        k = len(lists)
        q = PriorityQueue(maxsize=k)
        dummy = ListNode(None)
        curr = dummy
        for list_idx, node in enumerate(lists):
            if node: q.put((node.val, list_idx, node))
        while q.qsize() > 0:
            poped = q.get()
            curr.next, list_idx = poped[2], poped[1]
            curr = curr.next
            if curr.next: q.put((curr.next.val, list_idx, curr.next))
        return dummy.next

def createLinkedlist(lst):
    l = len(lst)
    if not l:
        return None

    head = ListNode(lst[0])
    cur = head
    for i in range(1, l):
        cur.next = ListNode(lst[i])
        i += 1
        cur = cur.next
    return head

def printList(head):
    while head:
        print(head.val, end='->')
        head = head.next
    print("\b\b  ")

if __name__ == "__main__":
    lists = [
        createLinkedlist([1,4,5]),
        createLinkedlist([1,3,4]),
        createLinkedlist([2,6])
    ]
    res = Solution().mergeKLists(lists)
    printList(res)



